semgrep-core/src/engine/Match_rules.ml

```diff
@@ -803,7 +755,79 @@ let rec (evaluate_formula : env -> R.formula -> range_with_mvars list) =
   | R.Not _ -> failwith "Invalid Not; you can only negate inside an And"
   | R.Leaf (R.MetavarCond _) ->
       failwith "Invalid MetavarCond; you can MetavarCond only inside an And"
- [@@profiling]
+
+and filter_ranges env xs cond =
+  xs
+  |> List.filter (fun r ->
+         let bindings = r.mvars in
+         match cond with
+         | R.CondGeneric e ->
+             let env = Eval_generic.bindings_to_env bindings in
+             Eval_generic.eval_bool env e
+         (* todo: would be nice to have CondRegexp also work on
+          * eval'ed bindings.
+          * We could also use re.match(), to be close to python, but really
+          * Eval_generic must do something special here with the metavariable
+          * which may not always be a string. The regexp is really done on
+          * the text representation of the metavar content.
+          *)
+         | R.CondPattern (mvar, formula) -> (
+             match List.assoc_opt mvar bindings with
+             | None ->
+                 Printf.eprintf "filter_ranges: CondPattern: %s not found" mvar;
+                 false
+             | Some mval -> (
+                 let mast = program_of_mvalue mval in
+                 let lazy_ast_and_errors = lazy (mast, []) in
+                 let xpatterns = xpatterns_in_formula formula in
```

Comment:
maybe a comment saying "the code below is quite similar to Match_rules.check below; we could maybe factorize things".